### ISSUE TITLE
Modified Command Set Files should be generated on the fly

### ISSUE THREAD

**Issue by latenitefilms** [fcpxhacks/fcpxhacks#117]
Currently, each time we add a new shortcut key to FCPX Hacks, we need to update the `defaultShortcutKeys()` function, as well as manually update all the various `.plist`, `.commandset` and `.strings` files for each language. We should really automate this, so that we can programmatically add new shortcuts to the various plist files.

Although, if we can eventually add a faster method for WRITING to plist files (i.e. via Objective-C), then maybe we can scrap the current method of adding our Hacks shortcuts to Final Cut Pro, and programatically write to the internal files, rather than "copy & replace".

In the meantime though - we should automate this.

**Comment by latenitefilms** [comment]
If you're up for it @randomeizer - this is something I could easily do, but you'll want to rewrite all my code after I've done it, so might be better if you just tackle from scratch! No stress or pressure though obviously, only if you can be bothered.

**Comment by latenitefilms** [comment]
Actually... rather than have a script that does this at "release" time - we should just generate these plist files "on the fly" when Hacks Shortcuts are enabled. This would mean we wouldn't have to manually edit these files anymore, which would be a massive help! It also means 3rd party plugins could also add things to the Final Cut Pro Command Editor which would be awesome.

I'll have a go at this later this week.

**Comment by latenitefilms** [comment]
FYI: To be tackled after #546 is merged into `develop`.

**Comment by latenitefilms** [comment]
@randomeizer - I've been thinking, if we're going to automate the Command Set creation, we might as well improve it whilst we're at it! What I want to do is use Actions instead of Commands within Final Cut Pro's Command Editor.

When you enable "Control CommandPost Shortcuts within Final Cut Pro", it will scan through all the available actions, assign them a UUID (with "cp" as a prefix - so we know what's a Final Cut Pro command versus a CommandPost command), then add all the actions to the Command Sets for all the available languages using the UUID as the identifier. It will also create a "mapping" JSON file that will be stored in the Application Support folder which allows you to translate between the UUID and the Action Parameters.

We'll create new `cp.commands` for every entry in the Command Set with the "cp" prefix, and we'll just make them trigger an action using the action parameters stored in the "mapping" file.

To simplify things, instead of worrying about automatically updating the Command Sets when a new version of CommandPost or Final Cut Pro comes out, we'll just add a "Rebuild Command Sets" button in the Preferences, and leave it up to the user to update their Command Sets when they update Final Cut Pro, CommandPost or add/remove a plugin/effect.

I believe this is all possible without major changes to the existing `cp.commands` or `core.action.manager` code - although I'm sure you'd eventually want to rewrite everything to make it cleaner.

I'd also love to use actions in CommandPost's own shortcut manager - but I have no idea what the user interface would be for that? I guess it could just be an empty table, and you have to "add" and "remove" things in it? Or we just do what we have with the MIDI Preferences, and just have an arbitrary amount of "slots" and you can assign whatever you want to one of those slots? That might be the simplest way to do it?

I'll try give this a bash over the next few days and/or weeks.

**Comment by randomeizer** [comment]
Yeah, it's been on my todo list for a while. My thought was to separate
"shortcuts" from commands and move them to a new "shortcuts manager", which
would trigger actions rather than commands.

With regards to a manual "rebuild", I'd rather we were smart enough to be
able to do it automatically, but I guess that could be a stop-gap option.

**Comment by latenitefilms** [comment]
> Yeah, it's been on my todo list for a while. My thought was to separate
> "shortcuts" from commands and move them to a new "shortcuts manager", which
> would trigger actions rather than commands.

Sounds like a good plan to me!

> With regards to a manual "rebuild", I'd rather we were smart enough to be
> able to do it automatically, but I guess that could be a stop-gap option.

The tricky thing is that you need admin permissions to replace the Command Set files with the Final Cut Pro bundle, which complicates things (i.e. #836). Having it user initiated makes the most sense - especially if we're going to add all actions into the Command Sets - otherwise the user will need to enter their admin password each time they add a Motion Template into Final Cut Pro. We could however have a notification at the top of the CommandPost menu, like we do with software updates, to alert users that rebuilding the command set is recommended?

I might still try have a bash at this later this week, unless you beat me to it.

**Comment by latenitefilms** [comment]
Aborted here: #1405
